group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+                // If not, assign to any other better suited member with the topic subscription.
+                if (!foundMatchingRackMember) {
+                    boolean isPartitionSticky = assignedStickyPartitions.contains(reassignablePartition);
+                    boolean isCurrentOwnerKnown = currentPartitionOwners.containsKey(reassignablePartition);
+
+                    if (rackInfo.useRackStrategy && !isPartitionSticky && isCurrentOwnerKnown) {
+                        String currentOwner = currentPartitionOwners.get(reassignablePartition);
+                        int currentMemberAssignmentSize = assignmentManager.targetAssignmentSize(member);
+                        int currentOwnerAssignmentSize = assignmentManager.targetAssignmentSize(currentOwner);
+
+                        // Check if reassignment is needed based on assignment sizes
+                        if (currentMemberAssignmentSize > currentOwnerAssignmentSize + 1) {
+                            reassignPartition(reassignablePartition, currentOwner);
+                            modified = true;
+                            reassignmentOccurred = true;
+                        }
+                    }
+
+                    // The topicIdPartition is already sticky and no other member with matching rack is better suited.
+                    if (!reassignmentOccurred) {
+                        for (String otherMember : membersPerTopic.get(reassignablePartition.topicId())) {
+                            if (assignmentManager.targetAssignmentSize(member) >
+                                assignmentManager.targetAssignmentSize(otherMember) + 1
+                            ) {
+                                reassignPartition(reassignablePartition);
+                                modified = true;
+                                reassignmentOccurred = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        } while (modified);
+    }
+
+    /**
+     * Reassigns a partition to an eligible member with the fewest current target assignments.
+     * <ul>
+     *   <li> Iterates over members sorted by ascending assignment size. </li>
+     *   <li> Selects the first member subscribed to the partition's topic. </li>
+     * </ul>
+     *
+     * @param partition         The partition to reassign.
+     * @throws AssertionError   If no subscribed member is found.
+     */
+    private void reassignPartition(TopicIdPartition partition) {
+        // Find the new member with the least assignment size.
+        String newOwner = null;
+        for (String anotherMember: sortedMembersByAssignmentSize) {
+            if (members.get(anotherMember).subscribedTopicIds().contains(partition.topicId())) {
+                newOwner = anotherMember;
+                break;
+            }
+        }
+
+        assert newOwner != null;
```

Comment:
Do we need this? It would be better to throw an IllegalStateException if we do. We usually don't use assert in the code base.

Reply:
ohh thanks for bringing it up, I wasn't aware of this, I was following the existing code from the client side assignor. Let me see what I can change it to

Reply:
I threw a PartitionAssignorException for now, I'm actually not fully convinced anymore about throwing an exception even though it made sense to me back then, I'll double check this and get back to you if that's okay. If you have any thoughts on this I would appreciate that as well! thanks!